pkg/handlers/archive.go

```diff
@@ -0,0 +1,159 @@
+package handlers
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+
+	"github.com/mholt/archiver/v4"
+	log "github.com/sirupsen/logrus"
+)
+
+var (
+	maxDepth = 5
+)
+
+// Archive is a handler for extracting and decompressing archives.
+type Archive struct {
+	newChunkData bytes.Buffer
+	maxSize      int
+	size         int
+}
+
+// New sets a default maximum size and current size counter.
+func (d *Archive) New() {
+	d.maxSize = 20 * 1024 * 1024 // 20MB
+	d.size = 0
+}
+
+// FromFile extracts the files from an archive.
+func (d *Archive) FromFile(data io.Reader) chan ([]byte) {
+	ctx := context.Background()
+	archiveChan := make(chan ([]byte), 512)
+	go func() {
+		defer close(archiveChan)
+		err := d.openArchive(ctx, 0, data, archiveChan)
+		if err != nil {
+			if errors.Is(err, archiver.ErrNoMatch) {
+				return
+			}
+			log.WithError(err).Debug("Error unarchiving chunk.")
+		}
+	}()
+	return archiveChan
+}
+
+// openArchive takes a reader and extracts the contents up to the maximum depth.
+func (d *Archive) openArchive(ctx context.Context, depth int, reader io.Reader, archiveChan chan ([]byte)) error {
+	if depth >= maxDepth {
+		return fmt.Errorf("max archive depth reached")
+	}
+	format, reader, err := archiver.Identify("", reader)
+	if err != nil {
+		if errors.Is(err, archiver.ErrNoMatch) && depth > 0 {
+			for {
+				chunk := make([]byte, 10*1024)
+				n, _ := reader.Read(chunk)
+				archiveChan <- chunk
+				if n < 512 {
+					break
+				}
+			}
+			fileBytes, err := d.ReadToMax(reader)
+			if err != nil {
+				return err
+			}
+			archiveChan <- fileBytes
+			return nil
+		}
+		return err
+	}
+	switch archive := format.(type) {
+	case archiver.Extractor:
+		err := archive.Extract(context.WithValue(ctx, "depth", depth+1), reader, nil, d.extractorHandler(archiveChan))
+		if err != nil {
+			return err
+		}
+		return nil
+	case archiver.Decompressor:
+		compReader, err := archive.OpenReader(reader)
+		if err != nil {
+			return err
+		}
+		fileBytes, err := d.ReadToMax(compReader)
+		if err != nil {
+			return err
+		}
+		newReader := bytes.NewReader(fileBytes)
+		return d.openArchive(ctx, depth+1, newReader, archiveChan)
+	}
+	return fmt.Errorf("Unknown archive type: %s", format.Name())
+}
+
+// IsFiletype returns true if the provided reader is an archive.
+func (d *Archive) IsFiletype(reader io.Reader) bool {
+	format, readerB, err := archiver.Identify("", reader)
```

Comment:
We must return readerB here to be used for further processing. The Handler interface must be updated and it should include a comment that a new reader must be returned.

>The returned io.Reader will always be non-nil and will read from the same point as the reader which was passed in; it should be used in place of the input stream after calling Identify() because it preserves and re-reads the bytes that were already read during the identification process.

https://pkg.go.dev/github.com/mholt/archiver/v4#Identify